cli_tools/common/image/importer/inflater.go

```diff
@@ -84,46 +105,79 @@ type inflaterFacade struct {
 	apiInflater   Inflater
 	daisyInflater Inflater
 	logger        logging.Logger
+	qemuChecksum  string
+	computeClient daisyCompute.Client
+	request       ImageImportRequest
 }
 
-func (facade *inflaterFacade) Inflate() (persistentDisk, shadowTestFields, error) {
+func (facade *inflaterFacade) Inflate() (persistentDisk, inflationInfo, error) {
 	var pd persistentDisk
-	var tf shadowTestFields
+	var ii inflationInfo
 	var err error
+	var fallbackReason string
 
-	//  Try with API inflater.
-	pd, tf, err = facade.apiInflater.Inflate()
-	if err == nil {
-		facade.logger.Metric(&pb.OutputInfo{
-			InflationType:   "api_success",
-			InflationTimeMs: []int64{tf.inflationTime.Milliseconds()},
-		})
-		return pd, tf, err
-	}
+	if facade.qemuChecksum == "" {
+		fallbackReason = "qemu_checksum_missing"
+	} else {
+		//  Run API inflater as the primary inflation method.
+		pd, ii, err = facade.apiInflater.Inflate()
 
-	if !isCausedByUnsupportedFormat(err) {
-		facade.logger.Metric(&pb.OutputInfo{
-			InflationType:   "api_failed",
-			InflationTimeMs: []int64{tf.inflationTime.Milliseconds()},
-		})
-		return pd, tf, err
+		if err != nil {
+			// If the inflation is failed but not due to unsupported format, don't rerun inflation.
+			if !isCausedByUnsupportedFormat(err) {
+				facade.logger.Metric(&pb.OutputInfo{
+					InflationType:   "api_failed",
+					InflationTimeMs: []int64{ii.inflationTime.Milliseconds()},
+				})
+				return pd, ii, err
+			}
+			fallbackReason = "unsupported_format"
+		}
+
+		if err == nil {
+			if isChecksumMatch(ii.checksum, facade.qemuChecksum) {
+				facade.logger.Metric(&pb.OutputInfo{
+					InflationType:   "api_success",
+					InflationTimeMs: []int64{ii.inflationTime.Milliseconds()},
+				})
+				return pd, ii, err
+			}
+
+			// If checksum mismatches , delete the corrupted disk.
+			facade.logger.User("Disk checksum mismatch, recreating...")
+			err = facade.computeClient.DeleteDisk(facade.request.Project, facade.request.Zone, getDiskName(facade.request.ExecutionID))
+			if err != nil {
+				return pd, ii, daisy.Errf("Tried to delete the disk after checksum mismatch is detected, but failed on: %v", err)
+			}
+			fallbackReason = "checksum_mismatch"
+		}
 	}
 
-	// Retry with daisy inflater.
-	pd, tf, err = facade.daisyInflater.Inflate()
+	// Now fallback to daisy inflater. As described above, it is due to one of the below reasons:
+	// 1. The API failed because it doesn't support the format of the image file.
+	// 2. Checksum mismatch, which means the API produced a corrupted disk.
+	// 3. QEMU checksum missed, which means we have no way to compare the checksum.
+	pd, ii, err = facade.fallbackToDaisyInflater(fallbackReason)
+	return pd, ii, err
```

Comment:
Can you just `return facade.fallbackToDaisyInflater(fallbackReason)`?

Reply:
done